server/tso/allocator_manager.go

```diff
@@ -0,0 +1,193 @@
+	client *clientv3.Client
+	// tso config
+	rootPath      string
+	saveInterval  time.Duration
+	maxResetTSGap func() time.Duration
+}
+
+// NewAllocatorManager creates a new TSO Allocator Manager.
+func NewAllocatorManager(etcd *embed.Etcd, client *clientv3.Client, rootPath string, saveInterval time.Duration, maxResetTSGap func() time.Duration) *AllocatorManager {
+	allocatorManager := &AllocatorManager{
+		allocatorGroups: make(map[string]*allocatorGroup),
+		etcd:            etcd,
+		client:          client,
+		rootPath:        rootPath,
+		saveInterval:    saveInterval,
+		maxResetTSGap:   maxResetTSGap,
+	}
+	return allocatorManager
+}
+
+func (am *AllocatorManager) getAllocatorPath(dcLocation string) string {
+	// For backward compatibility, the global timestamp's store path will still use the old one
+	if dcLocation == "global" {
+		return am.rootPath
+	}
+	return path.Join(am.rootPath, dcLocation)
+}
+
+// SetUpAllocator is used to set up an allocator, which will initialize the allocator and put it into allocator daemon.
+func (am *AllocatorManager) SetUpAllocator(parentCtx context.Context, parentCancel context.CancelFunc, dcLocation string, leadership *election.Leadership) error {
+	am.Lock()
+	defer am.Unlock()
+	switch dcLocation {
+	case "global":
+		am.allocatorGroups[dcLocation] = &allocatorGroup{
+			dcLocation:   dcLocation,
+			parentCtx:    parentCtx,
+			parentCancel: parentCancel,
+			leadership:   leadership,
+			allocator:    NewGlobalTSOAllocator(leadership, am.getAllocatorPath(dcLocation), am.saveInterval, am.maxResetTSGap),
+		}
+		if err := am.allocatorGroups[dcLocation].allocator.Initialize(); err != nil {
+			return errs.ErrSetAllocator.FastGenByArgs(err)
+		}
+		am.allocatorGroups[dcLocation].isInitialized = true
+	default:
+		// Todo: set up a Local TSO Allocator
+	}
+	return nil
+}
+
+// GetAllocator get the allocator by dc-location.
+func (am *AllocatorManager) GetAllocator(dcLocation string) (Allocator, error) {
+	am.RLock()
+	defer am.RUnlock()
+	allocatorGroup, exist := am.allocatorGroups[dcLocation]
+	if !exist {
+		return nil, errs.ErrGetAllocator.FastGenByArgs(fmt.Sprintf("%s allocator not found", dcLocation))
+	}
+	return allocatorGroup.allocator, nil
+}
+
+func (am *AllocatorManager) getAllocatorGroups() []*allocatorGroup {
+	am.RLock()
+	allocatorGroups := make([]*allocatorGroup, 0, len(am.allocatorGroups))
+	for _, ag := range am.allocatorGroups {
+		allocatorGroups = append(allocatorGroups, ag)
+	}
+	am.RUnlock()
+	return allocatorGroups
+}
+
+// AllocatorDaemon is used to update every allocator's TSO.
+func (am *AllocatorManager) AllocatorDaemon(serverCtx context.Context) {
+	tsTicker := time.NewTicker(UpdateTimestampStep)
+	defer tsTicker.Stop()
+
+	for {
+		select {
+		case <-tsTicker.C:
+			// Collect all dc-locations first
+			allocatorGroups := am.getAllocatorGroups()
+			// Update each allocator concurrently
+			for _, ag := range allocatorGroups {
+				go am.updateAllocator(ag)
+			}
```

Comment:
I mean we don't need sync.waitGroup and `go` anymore as `updateAllocator` functions are controlling by the mutex so they must be executed serially.

Reply:
Oh I get it!